examples/ex3/main.cpp

```diff
@@ -0,0 +1,280 @@
+#include "ConjGrad.hpp"
+#include "MCIntegrator.hpp"
+#include "FeedForwardNeuralNetwork.hpp"
+#include "PrintUtilities.hpp"
+
+
+//1D Target Function
+class Function1D {
+public:
+  //Function1D(){}
+  virtual ~Function1D(){}
+
+  // Function evaluation
+  virtual double f(const double &) = 0;
+  //                    ^input
+};
+
+// exp(-a*(x-b)^2)
+class Gaussian: public Function1D{
+private:
+  double _a;
+  double _b;
+
+public:
+  Gaussian(const double &a, const double &b): Function1D(){
+    _a = a;
+    _b = b;
+  }
+
+  double f(const double &x){
+    return exp(-_a*pow(x-_b, 2));
+  }
+};
+
+// Quadratic distance between NN and 1D target function
+class NNFitDistance1D: public MCIObservableFunctionInterface{
+private:
+  FeedForwardNeuralNetwork * _ffnn;
+  Function1D * _ftarget;
+public:
+  NNFitDistance1D(FeedForwardNeuralNetwork * ffnn, Function1D * ftarget): MCIObservableFunctionInterface(1, 1) {_ffnn = ffnn; _ftarget = ftarget;}
+protected:
+    void observableFunction(const double * in, double * out){
+
+      _ffnn->setInput(1, &in[0]);
+      _ffnn->FFPropagate();
+      out[0] = pow(_ffnn->getOutput(1) - _ftarget->f(in[0]), 2);
+
+    }
+};
+
+// Gradient of Quadratic distance between NN and 1D target function
+class NNFitGradient1D: public MCIObservableFunctionInterface{
+private:
+  FeedForwardNeuralNetwork * _ffnn;
+  Function1D * _ftarget;
+public:
+  NNFitGradient1D(FeedForwardNeuralNetwork * ffnn, Function1D * ftarget): MCIObservableFunctionInterface(1, ffnn->getNBeta()) {_ffnn = ffnn; _ftarget = ftarget;}
+protected:
+  void observableFunction(const double * in, double * out){
+
+    _ffnn->setInput(1, &in[0]);
+    _ffnn->FFPropagate();
+
+    double nnout = _ffnn->getOutput(1);
+    double fx = _ftarget->f(in[0]);
+
+    for(int i=0; i<_nobs; ++i){
+        out[i] = 2.*(nnout - fx)*_ffnn->getVariationalFirstDerivative(1, i);
+      }
+  }
+};
+
+//
+class FitNN1D: public NoisyFunctionWithGradient{
+private:
+  FeedForwardNeuralNetwork * _ffnn;
+  Function1D * _ftarget;
+  MCI * _mcif, * _mcig;
+  long _nmc;
+  double _x0, _step0;
+
+public:
+  FitNN1D(FeedForwardNeuralNetwork * ffnn, Function1D * ftarget, const long &nmc, double * irange): NoisyFunctionWithGradient(ffnn->getNBeta()){
+    _ffnn = ffnn;
+    _ftarget = ftarget;
+    _nmc = nmc;
+
+    _x0 = 0.5*(irange[0] + irange[1]);
+    _step0 = 0.1*(irange[1] - irange[0]);
+
+    _mcif = new MCI(1);
+    _mcig = new MCI(1);
+
+    MCIObservableFunctionInterface * fobs = new NNFitDistance1D(_ffnn, _ftarget);
+    MCIObservableFunctionInterface * gobs = new NNFitGradient1D(_ffnn, _ftarget);
+
+    _mcif->addObservable(fobs);
+    _mcig->addObservable(gobs);
+
+    double ** intrange = new double*[1];
+    intrange[0] = irange;
+    _mcif->setIRange(intrange);
+    _mcig->setIRange(intrange);
+
+    double targetacceptrate = 0.5;
+    _mcif->setTargetAcceptanceRate(&targetacceptrate);
+    _mcig->setTargetAcceptanceRate(&targetacceptrate);
+
+  }
+
```

Comment:
you didn't declare the destructor, where you de-allocate all the memory that you have had allocated!!!
_mcif, _mcig, fobs_ gobs

My suggestion is that wehenver you write the word "new", you should immediately write also the delete, without postponing!

Reply:
And btw, if you don't store the pointers fobs and gobs, how are you going to delete them at the end?
You should add them to the list of private class parameters